crates/serde-util/src/byte_count.rs

```diff
@@ -0,0 +1,353 @@
+        } else {
+            // raw u64 num of bytes
+            self.0.serialize(serializer)
+        }
+    }
+}
+
+impl<'de, const CAN_BE_ZERO: bool> Deserialize<'de> for ByteCount<CAN_BE_ZERO> {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        if deserializer.is_human_readable() {
+            deserializer.deserialize_any::<ByteCountVisitor<CAN_BE_ZERO>>(ByteCountVisitor)
+        } else {
+            deserializer.deserialize_u64::<ByteCountVisitor<CAN_BE_ZERO>>(ByteCountVisitor)
+        }
+    }
+}
+
+struct ByteCountVisitor<const CAN_BE_ZERO: bool>;
+
+// because we do what appears to be runtime-check fo the const CAN_BE_ZERO but
+// it's actually compile-time, compiler will inline the comparison but we need
+// to make clippy happy.
+#[allow(clippy::absurd_extreme_comparisons)]
+impl<'de, const CAN_BE_ZERO: bool> Visitor<'de> for ByteCountVisitor<CAN_BE_ZERO> {
+    type Value = ByteCount<CAN_BE_ZERO>;
+    fn expecting(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+        formatter.write_str("an integer or string")
+    }
+
+    // TOML Integer(i64) - toml doesn't support u64
+    fn visit_i64<E: de::Error>(self, value: i64) -> Result<Self::Value, E> {
+        if let Ok(val) = u64::try_from(value) {
+            if val <= 0 && !CAN_BE_ZERO {
+                return Err(E::invalid_value(
+                    de::Unexpected::Signed(value),
+                    &"non-zero value",
+                ));
+            }
+            Ok(ByteCount(val))
+        } else {
+            Err(E::invalid_value(
+                de::Unexpected::Signed(value),
+                &"i64 overflow",
+            ))
+        }
+    }
+
+    // JSON Integer(u64)
+    fn visit_u64<E: de::Error>(self, value: u64) -> Result<Self::Value, E> {
+        if value <= 0 && !CAN_BE_ZERO {
+            return Err(E::invalid_value(
+                de::Unexpected::Unsigned(value),
+                &"non-zero value",
+            ));
+        }
+        Ok(ByteCount(value))
+    }
+
+    fn visit_str<E: de::Error>(self, value: &str) -> Result<Self::Value, E> {
+        let val: Result<ByteSize, _> = value.parse();
+
+        if let Ok(val) = val {
+            if val.0 <= 0 && !CAN_BE_ZERO {
+                return Err(E::invalid_value(
+                    de::Unexpected::Str(value),
+                    &"non-zero value",
+                ));
+            }
+            Ok(ByteCount(val.0))
+        } else {
+            Err(E::invalid_value(
+                de::Unexpected::Str(value),
+                &"a human-readable byte count string",
+            ))
+        }
+    }
+}
+
+impl<const CAN_BE_ZERO: bool> SerializeAs<u64> for ByteCount<CAN_BE_ZERO> {
+    fn serialize_as<S>(source: &u64, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        Self(*source).serialize(serializer)
+    }
+}
```

Comment:
Shouldn't all `SerializeAs/DeserializeAs` implementations for zero-able types be only implemented for `ByteCount<true>`? Otherwise, when deserializing a 0 value using `ByteCount<false>`, it will probably crash.

Reply:
Ok, I see how you are using it in the options structs to enforce that a `u64` is actually non-zero when deserializing it. I am wondering whether we shouldn't change the type from u64 to `NonZeroU64` for those fields instead. The reason is that we can now construct a `CommonOptions` with 0 valued fields which we can serialize but fail to deserialize because of the non-zero constraint that is only enforced at deserialization time.